lib/block/blockupload.js

```diff
@@ -0,0 +1,199 @@
+
+        const source = new Asset(sourceUrl);
+        const target = new Asset(targetUrl, options.headers, uploadFile.multipartHeaders);
+        
+        if(!uploadFile.fileSize) {
+            const { size } = await getFileStats(uploadFile.filePath);
+            uploadFile.fileSize = size;
+        }
+
+        const contentType = options.headers && options.headers['content-type'];
+        const transferAsset = new TransferAsset(source, target, {
+            acceptRanges: true,
+            metadata: new AssetMetadata(uploadFile.filePath, contentType, uploadFile.fileSize)
+        });
+
+        const uploadURIs = uploadFile.fileUrl;
+        const minPartSize = uploadFile.minPartSize || 10;
+        const maxPartSize = uploadFile.maxPartSize;
+        
+        if(typeof uploadURIs === 'object' && Array.isArray(uploadURIs)) {
+            transferAsset.multipartTarget = new AssetMultipart(
+                uploadURIs,
+                minPartSize,
+                maxPartSize,
+                transferAsset.target.multipartHeaders
+            );
+        }
+
+        yield transferAsset;
+    }
+}
+
+class BlockUpload extends EventEmitter {
+
+    /**
+     * @typedef {Object} UploadFile
+     * @property {String} fileUrl AEM url where to upload the file
+     * @property {Number} fileSize Size of the file to upload
+     * @property {String} filePath Path on the local disk to upload
+     * @property {Blob} [blob] Browser blob to upload (instead of fileUrl)
+     * @property {Boolean} [createVersion=false] Create version on duplicates
+     * @property {String} [versionLabel] Version label to apply to the created/updated file
+     * @property {String} [versionComment] Version comment to apply to the created/updated file
+     * @property {Boolean} [replace=false] True if the existing file should be replaced
+     */
+    /**
+     * @typedef {Object} BlockUploadOptions
+     * @property {UploadFile[]} uploadFiles List of files that will be uploaded to the target URL. 
+     * @property {*} headers HTTP headers that will be included in each request sent to AEM
+     * @property {Boolean} concurrent If true, multiple files in the supplied list of upload files will transfer simultaneously. If false, only one file will transfer at a time, and the next file will not begin transferring until the current file finishes.
+     * @property {Number} maxConcurrent Maximum number of concurrent HTTP requests that are allowed
+     * @property {Number} [preferredPartSize] Preferred part size
+     */
+    /**
+     * Upload files to AEM
+     * 
+     * @param {BlockUploadOptions} options AEM upload options
+     */
+    async uploadFiles(options = {}) {
+        const preferredPartSize = options.preferredPartSize || DEFAULT_PART_SIZE;
+        const maxConcurrent = (options.concurrent && options.maxConcurrent) || DEFAULT_MAX_CONCURRENCY;
+        const retryOptions = {
+            retryMaxCount: options.retryCount || DEFAULT_RETRY_COUNT,
```

Comment:
I don't think `retryMaxCount` is needed here because I added the retry logic to the outer level in `uploadFileConcurrently` like we have for `uploadFile`: https://github.com/adobe/node-httptransfer/blob/9357ef6ed96e263cbefe3cd894cdd7b59abf5f0c/lib/file.js#L139

I think only `timeout` must be passed along to the fetch call options because it is needed to exit out of a fetch request when it's taking too long.

Reply:
`retryOptions` here will be used while transferring every part/block, so we would still want it. Given we have a retry for every part/block, having a retry at `uploadFileConcurrently` level would add even more retry.. (it may already crossover the [default retryMax](https://github.com/adobe/node-httptransfer/blob/9357ef6ed96e263cbefe3cd894cdd7b59abf5f0c/lib/retry.js#L39))

Reply:
I see thanks, I missed that the retry logic was built into transfer.js. 
The issue was we were not passing all the possible retry options here: https://github.com/adobe/node-httptransfer/blob/9357ef6ed96e263cbefe3cd894cdd7b59abf5f0c/lib/block/blockupload.js#L177
so some edge test cases with custom retry options like `replyWithAllErrors` were failing